libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -684,28 +684,33 @@ string YulUtilFunctions::overflowCheckedIntMulFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> product {
 				x := <cleanupFunction>(x)
 				y := <cleanupFunction>(y)
+				product := <cleanupFunction>(mul(x, y))
 				<?signed>
-					// overflow, if x > 0, y > 0 and x > (maxValue / y)
-					if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(<maxValue>, y))) { <panic>() }
-					// underflow, if x > 0, y < 0 and y < (minValue / x)
-					if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { <panic>() }
-					// underflow, if x < 0, y > 0 and x < (minValue / y)
-					if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { <panic>() }
-					// overflow, if x < 0, y < 0 and x < (maxValue / y)
-					if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { <panic>() }
+					// special case
+					if and(eq(x, sub(0, 1)), eq(y, <minValue>)) { <panic>() }
```

Comment:
So ``x == -1`` and ``y`` is the most negative value for the type... that means ``mul(x,y)`` is ``-<minValue>``, which after cleanup should be zero, shouldn't it?

But ``iszero(-1) and ``eq(<minValue>, sdiv(0, -1))`` are both false, so this will panic below already, won't it?

I'm probably missing something here, but given that this first special case seems redundant.

Reply:
> So `x == -1` and `y` is the most negative value for the type... that means `mul(x,y)` is `-<minValue>`, which after cleanup should be zero, shouldn't it?

Yes, that is how it works for types with less than 256 bits. But for 256 bits, `cleanup` will not change anything and  besides when we do `x * y = -1 * minValue`, the truncated result transforms back into minValue. Then when we do the division `product / x` in the check `eq(<minValue>, sdiv(0, -1))`, the quotient is truncated and once more turns into minValue.  
For simplicity, assume the max bit length is 16 instead of 256:
``` 
1111 1111 1111 1111		(X = -1)
1000 0000 0000 0000		(Y = minValue)

# P = X * Y
01000 0000 0000 0000		(P = -minValue) OVERFLOWN
1000 0000 0000 0000		(P = minValue) TRUNCATED 

# Q = P / X 
01000 0000 0000 0000		(Q = -minValue) OVERFLOWN
1000 0000 0000 0000		(Q = minValue) TRUNCATED 
```
Thus, both `iszero(-1)` and `eq(<minValue>, sdiv(minValue, -1))` are going to be true and the overflow will not be detected properly. 

It is unfortunate to have this special case for only one int type. I haven't been able to figure out a way of removing it though :(
We could have an `if` separating 256bit from other types, but when I tested this before, the benchmarks were marginally worse.

Reply:
You can just pass a whiskers boolean into the template to separate the 256-bit case from the <256-bit cases - just like ``<?signed>`` works. That will come without runtime cost at all and all <256-bit cases should get cheaper without the additional check.

Reply:
Yeah, that's what I did... maybe I read it wrong and though it was worse. Anyway, I am going to test that and compare.